server.go

```diff
@@ -319,21 +321,45 @@ func (srv *Server) ListenAndServe() error {
 }
 
 func (srv *Server) initServeUDP(connUDP *coapNet.ConnUDP) error {
-	return srv.serveUDP(newShutdownWithContext(srv.doneChan), connUDP)
+	srv.doneLock.Lock()
+	if srv.doneChan != nil {
+		return fmt.Errorf("server already serve connections")
+	}
+	doneChan := make(chan struct{})
+	srv.doneChan = doneChan
+	srv.doneLock.Unlock()
+
+	return srv.serveUDP(newShutdownWithContext(doneChan), connUDP)
 }
 
 func (srv *Server) initServeTCP(conn *coapNet.Conn) error {
+	srv.doneLock.Lock()
+	if srv.doneChan != nil {
+		return fmt.Errorf("server already serve connections")
+	}
+	doneChan := make(chan struct{})
+	srv.doneChan = doneChan
+	srv.doneLock.Unlock()
+
 	if srv.NotifyStartedFunc != nil {
 		srv.NotifyStartedFunc()
 	}
-	return srv.serveTCPConnection(newShutdownWithContext(srv.doneChan), conn)
+	return srv.serveTCPConnection(newShutdownWithContext(doneChan), conn)
 }
 
 func (srv *Server) initServeDTLS(conn *coapNet.Conn) error {
+	srv.doneLock.Lock()
+	if srv.doneChan != nil {
+		return fmt.Errorf("server already serve connections")
```

Comment:
Please could you always use `defer` for unlocking? This path does not unlock. Consider using [Event](https://godoc.org/google.golang.org/grpc/internal/grpcsync).